### DISCUSSION TITLE
Port forward for connections without shell access

### DISCUSSION THREAD

**Discussion by fullc0ntr0l** [ronf/asyncssh#604] [Q&A]
What's the equivalent of ssh `-N` argument when doing local port forwarding with no shell access user?
```bash
ssh -N -L 8080:xxxxx:80 user@ssh-server
```

I'm trying to run the first example from [here](https://asyncssh.readthedocs.io/en/latest/#port-forwarding) but it closes my connection

```bash
DEBUG:asyncssh:[conn=0] Trying password auth
INFO:asyncssh:[conn=0] Auth for user user succeeded
INFO:asyncssh:[conn=0] Creating local TCP forwarder from dynamic port to xxxxx, port 80
DEBUG:asyncssh:[conn=0] Assigning dynamic port 44119
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
INFO:asyncssh:[conn=0] Connection closed
```

**Comment by fullc0ntr0l** [comment]
Actually that happens on user with shell access too. Here's the code I'm trying to execute

```py
import asyncio
import asyncssh
import sys
import logging

logging.basicConfig(level="DEBUG")
asyncssh.set_debug_level(2)


async def run_client() -> None:
    async with asyncssh.connect(
        HOST,
        username=USERNAME,
        password=PASSWORD,
    ) as conn:
        listener = await conn.forward_local_port(
            "",
            0,
            REMOTE_ADDR,
            REMOTE_PORT,
        )
        await listener.wait_closed()


try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit("SSH connection failed: " + str(exc))
```

**Reply by fullc0ntr0l** [reply]
Looks like if I replace `listener.wait_closed()` with `conn.wait_closed()` it is not closing the connection and runs forever as it was intended. Is this missing in the examples?

**Reply by ronf** [reply]
If you really want the equivalent of "ssh -N -L 8080:xxxxx:80 user@ssh-server", you'll want to specify 8080 as the port number in the `conn.forward_local_port()` rather than using 0, which means to assign a dynamic port. Otherwise, though, this looks right.

As for needing to use `conn.wait_closed()` instead of `listener.wait_closed()`, that doesn't make sense to me. Unless you have something explicitly calling `listener.close()`, the two should be equivalent. I just tried the code you pasted above, and it is working for me, accepting multiple connections and running indefinitely.

**Reply by fullc0ntr0l** [reply]
That's weird, so listener.wait_closed is closing my connection whatever I'm using, local port or unix socket, and conn.wait_closed is keeping it alive. What version of python have you tested that? I'm on 3.12.0

**Reply by ronf** [reply]
Interesting. I'm running the AsyncSSH unit tests on Python 3.12 without a problem, but I am able to reproduce the problem of `listener.wait_closed()` not blocking like it should. but it working with `conn.wait_closed()`. I'll take a closer look. Both versions work fine for me on Python 3.11.

**Reply by fullc0ntr0l** [reply]
Could be related to [this issue](https://github.com/python/cpython/issues/79033) and fixed in next python version
> [gh-79033](https://github.com/python/cpython/issues/79033): Another attempt at fixing [asyncio.Server.wait_closed()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server.wait_closed). It now blocks until both conditions are true: the server is closed, and there are no more active connections. (This means that in some cases where in 3.12.0 this function would incorrectly have returned immediately, it will now block; in particular, when there are no active connections but the server hasn’t been closed yet.)

**Reply by ronf** [reply]
Yes - it looks like it. I tracked it down so far to a change in `Server.wait_closed()` from:

```python
        if self._sockets is None or self._waiters is None:
```
to:
```python
        if self._waiters is None or self._active_count == 0:
```
If I add a sleep before calling wait_closed() to give time to successfully open an active socket on the listener, things work as they did before. However, with an active count of 0, it never tries to wait.

**Reply by ronf** [reply]
The following change in Python 3.12 seems to work (included in https://github.com/python/cpython/pull/111336):

```patch
--- /tmp/base_events.py.orig	2023-11-04 09:16:12
+++ base_events.py	2023-11-04 09:16:37
@@ -377,7 +377,7 @@
             self._serving_forever_fut = None

     async def wait_closed(self):
-        if self._waiters is None or self._active_count == 0:
+        if self._waiters is None:
             return
         waiter = self._loop.create_future()
         self._waiters.append(waiter)
```
